File: eval-reviews/2026-09-12-chained/workflow-recovery/python/astra/r1/starter/leased.py
"""Leased execution with durable acquisitions and separately delivered responses."""
from dataclasses import dataclass

from workflow import (
    TIME_LIMIT, RunState, Simulator, StepState, Workflow, identifier, integer,
    object_fields, parse_command, parse_step, require, validate_graph,
)


def parse_leased(raw):
    obj = object_fields(raw, {"steps", "commands", "workers"},
                        {"max_attempts", "retry_delay", "lease_duration"})
    require(type(obj["workers"]) is list and bool(obj["workers"]))
    workers = tuple(identifier(worker) for worker in obj["workers"])
    require(len(workers) == len(set(workers)))
    require(type(obj["steps"]) is list and bool(obj["steps"]))
    require(type(obj["commands"]) is list and len(obj["commands"]) <= 2000)
    steps = tuple(parse_step(step) for step in obj["steps"])
    for cmd in obj["commands"]:
        require(type(cmd) is dict and type(cmd.get("op")) is str)
        op = cmd["op"]
        if op in ("start", "cancel", "advance", "observe"):
            parse_command(cmd)
        elif op in ("crash", "restart", "claim", "renew", "call"):
            fields = {"op", "worker"}
            if op in ("renew", "call"):
                fields.add("ticket")
            object_fields(cmd, fields)
            identifier(cmd["worker"])
            if "ticket" in cmd:
                integer(cmd["ticket"], 1, TIME_LIMIT)
        else:
            require(op == "deliver")
            object_fields(cmd, {"op", "ticket"})
            integer(cmd["ticket"], 1, TIME_LIMIT)
    workflow = Workflow(steps, (), integer(obj.get("max_attempts", 3), 1, 10),
                        integer(obj.get("retry_delay", 2), 1, 1_000_000))
    duration = integer(obj.get("lease_duration", 5), 1, 1_000_000)
    validate_graph(steps)
    return workflow, workers, duration, obj["commands"]


@dataclass
class LeasedStep(StepState):
    lease: dict | None = None


@dataclass
class Ticket:
    worker: str
    run: RunState
    step: LeasedStep
    definition: object
    kind: str
    receipt: dict | None = None
    delivered: bool = False


class LeasedSimulator(Simulator):
    def __init__(self, raw):
        workflow, workers, self.duration, self.commands = parse_leased(raw)
        super().__init__(workflow)
        self.workers = dict.fromkeys(workers, True)
        self.tickets = {}
        self.audit = []

    def live(self, number, ticket):
        lease = ticket.step.lease
        return (lease is not None and lease["ticket"] == number
                and self.now < lease["expires"])

    def deadline(self, delay):
        result = self.now + delay
        require(result <= TIME_LIMIT, "TIME_OVERFLOW")
        return result

    def claim(self, worker):
        require(not any(step.lease is not None and step.lease["worker"] == worker
                        and self.now < step.lease["expires"]
                        for run in self.runs for step in run.steps), "WORKER_BUSY")
        action = None
        for run in self.runs:
            for step, definition in zip(run.steps, self.workflow.steps):
                if step.status == "running" and step.lease["expires"] <= self.now:
                    action = run, step, definition
                    break
            if action is not None:
                break
        if action is None:
            for run in self.runs:
                if run.status != "active":
                    continue
                succeeded = {step.id for step in run.steps if step.status == "succeeded"}
                for step, definition in zip(run.steps, self.workflow.steps):
                    if (step.status == "pending" and step.ready_at <= self.now
                            and all(dep in succeeded for dep in definition.needs)):
                        action = run, step, definition
                        break
                if action is not None:
                    break
        if action is None:
            return {"ticket": None}
        expires = self.deadline(self.duration)
        run, step, definition = action
        number = len(self.tickets) + 1
        if step.status == "pending":
            step.status = "running"
            step.attempts += 1
        step.lease = {"worker": worker, "ticket": number, "expires": expires}
        self.tickets[number] = Ticket(worker, run, step, definition,
                                      "execute" if run.status == "active" else "lookup")
        return {"ticket": number}

    def call(self, number, ticket):
        if not self.live(number, ticket):
            return {"outcome": "stale"}
        if ticket.receipt is None:
            key = (ticket.run.id, ticket.step.id)
            attempt = ticket.step.attempts
            if ticket.kind == "lookup":
                outcome = self.service.lookup(key, attempt)
            else:
                outcome = self.service.execute(key, ticket.definition.amount, attempt,
                                               ticket.definition.failures)
            ticket.receipt = {"kind": ticket.kind, "outcome": outcome}
            self.audit.append({"worker": ticket.worker, "ticket": number,
                               "kind": ticket.kind, "key": list(key),
                               "attempt": attempt, "outcome": outcome})
        return dict(ticket.receipt)

    def expire_running(self, run):
        for step in run.steps:
            if step.status == "running":
                step.lease["expires"] = self.now

    def deliver(self, number, ticket):
        if (ticket.receipt is None or ticket.delivered or not self.live(number, ticket)
                or not self.workers[ticket.worker]):
            return {"committed": False}
        run, step = ticket.run, ticket.step
        outcome = ticket.receipt["outcome"]
        if ticket.kind == "lookup":
            step.status = ("succeeded" if outcome == "found" else
                           "cancelled" if run.status == "cancelling" else "blocked")
        elif outcome == "transient":
            if step.attempts < self.workflow.max_attempts:
                step.ready_at = self.deadline(self.workflow.retry_delay)
                step.status = "pending"
            else:
                step.status = "failed"
                for other in run.steps:
                    if other.status == "pending":
                        other.status = "blocked"
                run.status = "failing"
                self.expire_running(run)
        else:
            step.status = "succeeded"
        step.lease = None
        ticket.delivered = True
        if not any(other.status == "running" for other in run.steps):
            if run.status == "cancelling":
                run.status = "cancelled"
            elif run.status == "failing":
                run.status = "failed"
        if run.status == "active" and all(other.status == "succeeded" for other in run.steps):
            run.status = "succeeded"
        return {"committed": True}

    def cancel(self, run_id):
        run = next((run for run in self.runs if run.id == run_id), None)
        require(run is not None, "UNKNOWN_RUN")
        if run.status != "active":
            return
        run.cancel_requested = True
        for step in run.steps:
            if step.status == "pending":
                step.status = "cancelled"
        self.expire_running(run)
        run.status = ("cancelling" if any(step.status == "running" for step in run.steps)
                      else "cancelled")

    def apply(self, cmd):
        op = cmd["op"]
        if "worker" in cmd:
            worker = cmd["worker"]
            require(worker in self.workers, "UNKNOWN_WORKER")
            if op == "restart":
                require(not self.workers[worker], "WORKER_UP")
            else:
                require(self.workers[worker], "WORKER_DOWN")
        if "ticket" in cmd:
            number = cmd["ticket"]
            require(number in self.tickets, "UNKNOWN_TICKET")
            ticket = self.tickets[number]
            if "worker" in cmd:
                require(ticket.worker == worker, "WRONG_WORKER")
        if op == "start":
            require(all(run.id != cmd["run"] for run in self.runs), "DUPLICATE_RUN")
            self.runs.append(RunState(cmd["run"], [LeasedStep(step.id, ready_at=self.now)
                                                  for step in self.workflow.steps]))
        elif op == "advance":
            self.now = self.deadline(cmd["by"])
        elif op == "observe":
            return self.snapshot()
        elif op in ("crash", "restart"):
            self.workers[worker] = op == "restart"
        elif op == "claim":
            return self.claim(worker)
        elif op == "renew":
            live = self.live(number, ticket)
            if live:
                ticket.step.lease["expires"] = self.deadline(self.duration)
            return {"renewed": live}
        elif op == "call":
            return self.call(number, ticket)
        elif op == "deliver":
            return self.deliver(number, ticket)
        elif op == "cancel":
            self.cancel(cmd["run"])
        return None

    def snapshot(self):
        snapshot = super().snapshot()
        del snapshot["up"]
        snapshot["workers"] = [{"id": worker, "up": up} for worker, up in self.workers.items()]
        for saved, run in zip(snapshot["runs"], self.runs):
            for value, step in zip(saved["steps"], run.steps):
                value["lease"] = dict(step.lease) if step.lease is not None else None
        return snapshot

    def run(self):
        results = [self.apply(cmd) for cmd in self.commands]
        final = self.snapshot()
        final["calls"] = self.audit
        final["effects"] = [{"key": list(effect.key), "amount": effect.amount}
                            for effect in self.service.effects]
        return {"results": results, "final": final}

File: eval-reviews/2026-09-12-chained/workflow-recovery/typescript/astra/r1/starter/leased.ts
import {
  DomainError, TIME_LIMIT, fields, identifier, integer, require,
  parseStep, validateGraph, MockService,
} from "./workflow.ts";
import type { StepDefinition, StepState, RunState } from "./workflow.ts";

type Command =
  | { op: "start" | "cancel"; run: string }
  | { op: "advance"; by: number }
  | { op: "observe" }
  | { op: "crash" | "restart" | "claim"; worker: string }
  | { op: "renew" | "call"; worker: string; ticket: number }
  | { op: "deliver"; ticket: number };
interface Workflow {
  steps: StepDefinition[];
  commands: Command[];
  workers: string[];
  maxAttempts: number;
  retryDelay: number;
  leaseDuration: number;
}
function parseCommand(raw: unknown): Command {
  require(raw !== null && typeof raw === "object" && !Array.isArray(raw));
  const obj = raw as Record<string, unknown>;
  switch (obj.op) {
    case "start": case "cancel":
      fields(raw, ["op", "run"]);
      return { op: obj.op, run: identifier(obj.run) };
    case "advance":
      fields(raw, ["op", "by"]);
      return { op: obj.op, by: integer(obj.by, 0, TIME_LIMIT) };
    case "observe":
      fields(raw, ["op"]);
      return { op: obj.op };
    case "crash": case "restart": case "claim":
      fields(raw, ["op", "worker"]);
      return { op: obj.op, worker: identifier(obj.worker) };
    case "renew": case "call":
      fields(raw, ["op", "worker", "ticket"]);
      return { op: obj.op, worker: identifier(obj.worker), ticket: integer(obj.ticket, 1, TIME_LIMIT) };
    case "deliver":
      fields(raw, ["op", "ticket"]);
      return { op: obj.op, ticket: integer(obj.ticket, 1, TIME_LIMIT) };
    default: throw new DomainError("INVALID_INPUT");
  }
}
export function parseLeasedWorkflow(raw: unknown): Workflow {
  const obj = fields(raw, ["steps", "commands", "workers"], ["max_attempts", "retry_delay", "lease_duration"]);
  require(Array.isArray(obj.steps) && obj.steps.length > 0);
  require(Array.isArray(obj.commands) && obj.commands.length <= 2000);
  require(Array.isArray(obj.workers) && obj.workers.length > 0);
  const workers = obj.workers.map(identifier);
  require(new Set(workers).size === workers.length);
  const workflow: Workflow = {
    steps: obj.steps.map(parseStep), commands: obj.commands.map(parseCommand), workers,
    maxAttempts: integer(Object.hasOwn(obj, "max_attempts") ? obj.max_attempts : 3, 1, 10),
    retryDelay: integer(Object.hasOwn(obj, "retry_delay") ? obj.retry_delay : 2, 1, 1_000_000),
    leaseDuration: integer(Object.hasOwn(obj, "lease_duration") ? obj.lease_duration : 5, 1, 1_000_000),
  };
  validateGraph(workflow.steps);
  return workflow;
}
interface Lease { worker: string; ticket: number; expires: number }
interface Step extends StepState { lease: Lease | null }
interface Run extends Omit<RunState, "status" | "steps"> {
  status: RunState["status"] | "failing";
  steps: Step[];
}
type Receipt =
  | { kind: "execute"; outcome: "applied" | "replayed" | "transient" }
  | { kind: "lookup"; outcome: "found" | "missing" };
interface Ticket {
  id: number;
  worker: string;
  run: Run;
  step: Step;
  definition: StepDefinition;
  attempt: number;
  kind: Receipt["kind"];
  receipt?: Receipt;
  delivered: boolean;
}

/** Lease acquisitions and transport receipts survive independently of workers. */
export class LeasedSimulator {
  private now = 0;
  private readonly workflow: Workflow;
  private readonly workers: { id: string; up: boolean }[];
  private readonly runs: Run[] = [];
  private readonly tickets = new Map<number, Ticket>();
  private readonly service = new MockService();
  private readonly calls: (MockService["calls"][number] & { worker: string; ticket: number })[] = [];
  constructor(workflow: Workflow) {
    this.workflow = workflow;
    this.workers = workflow.workers.map(id => ({ id, up: true }));
  }
  private deadline(delay: number): number {
    const deadline = this.now + delay;
    require(deadline <= TIME_LIMIT, "TIME_OVERFLOW");
    return deadline;
  }
  private ticket(id: number): Ticket {
    const ticket = this.tickets.get(id);
    require(ticket, "UNKNOWN_TICKET");
    return ticket;
  }
  private live(ticket: Ticket): boolean {
    const lease = ticket.step.lease;
    return ticket.step.status === "running" && lease !== null &&
      lease.ticket === ticket.id && lease.worker === ticket.worker && this.now < lease.expires;
  }
  private select(): { run: Run; step: Step; definition: StepDefinition } | undefined {
    for (const run of this.runs) {
      const index = run.steps.findIndex(step => step.status === "running" &&
        step.lease !== null && step.lease.expires <= this.now);
      if (index >= 0) return { run, step: run.steps[index]!, definition: this.workflow.steps[index]! };
    }
    for (const run of this.runs) {
      if (run.status !== "active") continue;
      const index = run.steps.findIndex((step, i) => step.status === "pending" &&
        step.ready_at <= this.now && this.workflow.steps[i]!.needs.every(dep =>
          run.steps.some(other => other.id === dep && other.status === "succeeded")));
      if (index >= 0) return { run, step: run.steps[index]!, definition: this.workflow.steps[index]! };
    }
    return undefined;
  }
  private claim(worker: string): { ticket: number | null } {
    require(!this.runs.some(run => run.steps.some(step => step.lease?.worker === worker &&
      this.now < step.lease.expires)), "WORKER_BUSY");
    const action = this.select();
    if (!action) return { ticket: null };
    const expires = this.deadline(this.workflow.leaseDuration);
    const id = this.tickets.size + 1;
    const { run, step } = action;
    if (step.status === "pending") {
      step.status = "running";
      step.attempts++;
    }
    step.lease = { worker, ticket: id, expires };
    this.tickets.set(id, {
      ...action, id, worker, attempt: step.attempts,
      kind: run.status === "active" ? "execute" : "lookup", delivered: false,
    });
    return { ticket: id };
  }
  private call(ticket: Ticket): Receipt | { outcome: "stale" } {
    if (!this.live(ticket)) return { outcome: "stale" };
    if (!ticket.receipt) {
      const key: [string, string] = [ticket.run.id, ticket.step.id];
      ticket.receipt = ticket.kind === "execute"
        ? { kind: "execute", outcome: this.service.execute(key, ticket.definition.amount,
            ticket.attempt, ticket.definition.failures) }
        : { kind: "lookup", outcome: this.service.lookup(key, ticket.attempt) };
      this.calls.push({ worker: ticket.worker, ticket: ticket.id, key,
        attempt: ticket.attempt, ...ticket.receipt });
    }
    return { ...ticket.receipt };
  }
  /** Revocation leaves the old lease visible, but fences all its messages. */
  private expireRunning(run: Run): void {
    for (const step of run.steps)
      if (step.status === "running" && step.lease) step.lease.expires = this.now;
  }
  private deliver(ticket: Ticket): { committed: boolean } {
    if (ticket.delivered || !ticket.receipt || !this.live(ticket) ||
        !this.workers.find(worker => worker.id === ticket.worker)!.up)
      return { committed: false };
    const { step, run, receipt } = ticket;
    if (receipt.kind === "lookup") {
      step.status = receipt.outcome === "found" ? "succeeded"
        : run.status === "cancelling" ? "cancelled" : "blocked";
    } else if (receipt.outcome === "transient") {
      if (step.attempts < this.workflow.maxAttempts) {
        step.ready_at = this.deadline(this.workflow.retryDelay);
        step.status = "pending";
      } else {
        step.status = "failed";
        for (const other of run.steps)
          if (other.status === "pending") other.status = "blocked";
        run.status = "failing";
        this.expireRunning(run);
      }
    } else {
      step.status = "succeeded";
    }
    step.lease = null;
    ticket.delivered = true;
    if (!run.steps.some(other => other.status === "running")) {
      if (run.status === "cancelling") run.status = "cancelled";
      else if (run.status === "failing") run.status = "failed";
      else if (run.steps.every(other => other.status === "succeeded")) run.status = "succeeded";
    }
    return { committed: true };
  }
  private apply(command: Command): unknown {
    // Availability errors precede ticket validation, even for stale tickets.
    if ("worker" in command) {
      const worker = this.workers.find(worker => worker.id === command.worker);
      require(worker, "UNKNOWN_WORKER");
      if (command.op === "restart") {
        require(!worker.up, "WORKER_UP");
        worker.up = true;
        return null;
      }
      require(worker.up, "WORKER_DOWN");
      if (command.op === "crash") { worker.up = false; return null; }
      if (command.op === "claim") return this.claim(worker.id);
      if ("ticket" in command) {
        const ticket = this.ticket(command.ticket);
        require(ticket.worker === worker.id, "WRONG_WORKER");
        if (command.op === "call") return this.call(ticket);
        if (!this.live(ticket)) return { renewed: false };
        ticket.step.lease!.expires = this.deadline(this.workflow.leaseDuration);
        return { renewed: true };
      }
    }
    switch (command.op) {
      case "start":
        require(!this.runs.some(run => run.id === command.run), "DUPLICATE_RUN");
        this.runs.push({ id: command.run, status: "active", cancel_requested: false,
          steps: this.workflow.steps.map(step => ({ id: step.id, status: "pending",
            attempts: 0, ready_at: this.now, lease: null })) });
        return null;
      case "advance": this.now = this.deadline(command.by); return null;
      case "observe": return this.snapshot();
      case "deliver": return this.deliver(this.ticket(command.ticket));
      case "cancel": {
        const run = this.runs.find(run => run.id === command.run);
        require(run, "UNKNOWN_RUN");
        if (run.status !== "active") return null;
        run.cancel_requested = true;
        for (const step of run.steps)
          if (step.status === "pending") step.status = "cancelled";
        this.expireRunning(run);
        run.status = run.steps.some(step => step.status === "running") ? "cancelling" : "cancelled";
        return null;
      }
    }
  }
  private snapshot() {
    return { now: this.now, workers: this.workers.map(worker => ({ ...worker })),
      runs: this.runs.map(run => ({ ...run, steps: run.steps.map(step => ({ ...step,
        lease: step.lease ? { ...step.lease } : null })) })) };
  }
  run() {
    const results = this.workflow.commands.map(command => this.apply(command));
    return { results, final: { ...this.snapshot(), calls: this.calls, effects: this.service.effects } };
  }
}
